### PR TITLE
Index checker escape package

### PR DESCRIPTION
**Pull request by NhatDinh**
This pull request is based on [PR#10](https://github.com/typetools/guava/pull/10)

### CODE REVIEW

**Comment by NhatDinh** [comment]
Please review and let me know if any changes are needed.
@kelloggm @panacekcz

File: guava/src/com/google/common/escape/UnicodeEscaper.java [hunk 1]
```diff
@@ -154,13 +157,14 @@ protected int nextEscapeIndex(CharSequence csq, int start, int end) {
...
-  protected final String escapeSlow(String s, int index) {
+  @SuppressWarnings(value = {"assignment.type.incompatible", "argument.type.incompatible"})//for now
+  protected final String escapeSlow(String s, @IndexOrHigh("#1") int index) {
     int end = s.length();
 
     // Get a destination buffer and setup some loop variables.
     char[] dest = Platform.charBufferFromThreadLocal();
-    int destIndex = 0;
-    int unescapedChunkStart = 0;
+    @NonNegative @LTLengthOf(value = "dest", offset = "(index - unescapedChunkStart) - 1") int destIndex = 0;
```
Comment (panacekcz) [inline guava/src/com/google/common/escape/UnicodeEscaper.java]: A I have commented in a google doc:
The internal expression parser is currently limited and cannot parse expressions with parentheses like this. In this case, they are superfluous anyway.
Moreover, you cannot refer to a variable before it is declared

Comment (NhatDinh) [reply guava/src/com/google/common/escape/UnicodeEscaper.java]: I'll fix it right away.

File: guava/src/com/google/common/escape/ArrayBasedCharEscaper.java [hunk 2]
```diff
@@ -18,6 +18,9 @@
 
 import com.google.common.annotations.Beta;
 import com.google.common.annotations.GwtCompatible;
+import org.checkerframework.checker.index.qual.LTEqLengthOf;
+import org.checkerframework.checker.index.qual.LengthOf;
+
```
Comment (panacekcz) [inline guava/src/com/google/common/escape/ArrayBasedCharEscaper.java]: In guava, imports should be formatted according to the [google style guide](https://google.github.io/styleguide/javaguide.html#s3.3-import-statements) (don't use `*`, sort alphabetically, don't insert blank lines).

File: guava/src/com/google/common/escape/UnicodeEscaper.java [hunk 3]
```diff
@@ -127,7 +130,7 @@ public String escape(String string) {
    * @throws IllegalArgumentException if the scanned sub-sequence of {@code csq} contains invalid
    *     surrogate pairs
    */
-  protected int nextEscapeIndex(CharSequence csq, int start, int end) {
+  protected @NonNegative int nextEscapeIndex(CharSequence csq, @IndexFor("#1") int start, @IndexFor("#1") int end) {
```
Comment (panacekcz) [inline guava/src/com/google/common/escape/UnicodeEscaper.java]: Seems like there are some suppressed warnings related to this method, so let's look at it.

First, the documentation says that `end` is an index immediately after a character (in `csq`). Right from that, we can expect that `end` can be equal to `csq.length()` (an index after the last character). Therefore, `@IndexFor("#1")` is too strong for `end`, because it doesn't allow it to be `csq.length()`. To allow that, use `@IndexOrHigh("#1")`.

Second, the documentation states that `start` is an index of a character. That seems to suggest, that `@IndexFor` is the right annotation for it. However, when a method takes a sequence and two indices that specify a subsequence within that sequence, usually empty sequences are also allowed.
(As an example of that, consider `String.substring`. These calls `"abc".substring(0, 0)`, `"abc".substring(0, 3)`, `"abc".substring(3, 3)`, and `"".substring(0, 0)` are all valid.)
Is it the same case here? When the documentation does not provide a clear answer (often the case with non-public methods), it's helpful to look at the implementation and usage of the method, to see what is the intended behavior.
- The method is called from `UnicodeEscaper.escape(String)`.  Users can call `escape("")`, which will in turn call `nextEscapeIndex("", 0, 0)`. From that, we can conclude that `nextEscapeIndex` is supposed to work correctly even if `start == csq.length()`. Therefore, I would argue that `start` should also be annotated `@IndexOrHigh("#1")`.
- Let's look what happens in the implementation. When both indices are the same, the while loop is skipped and the method simply returns `start`. So it will behave correctly.

Last, the method returns an index. It would be good to express that with an annotation. Clearly, the returned value can be equal to the length of the sequence. So again, `@IndexOrHigh("#1")` is the right annotation for the return type.

File: guava/src/com/google/common/escape/ArrayBasedCharEscaper.java (hunk 2 above)
Comment (panacekcz) [reply guava/src/com/google/common/escape/ArrayBasedCharEscaper.java]: That being said, this has low priority and can be easily fixed when everything else is ready.

File: guava/src/com/google/common/escape/UnicodeEscaper.java (hunk 1 above)
Comment (panacekcz) [reply guava/src/com/google/common/escape/UnicodeEscaper.java]: @NhatDinh, is there any problem with this?

File: guava/src/com/google/common/escape/UnicodeEscaper.java [hunk 4]
```diff
@@ -242,7 +246,7 @@ protected final String escapeSlow(String s, int index) {
    * @return the Unicode code point for the given index or the negated value of the trailing high
    *     surrogate character at the end of the sequence
    */
-  protected static int codePointAt(CharSequence seq, int index, int end) {
+  protected static int codePointAt(CharSequence seq, @IndexOrHigh("#1") int index, @IndexFor("#1") int end) {
```
Comment (panacekcz) [inline guava/src/com/google/common/escape/UnicodeEscaper.java]: @NhatDinh, why did you change the annotation of `index` to `@IndexOrHigh`, contrary to my suggestion?

Comment (NhatDinh) [reply guava/src/com/google/common/escape/UnicodeEscaper.java]: @panacekcz it was an accident.

Comment (NhatDinh) [reply guava/src/com/google/common/escape/UnicodeEscaper.java]: @panacekcz after taking another look, I think `int end` should be annotated as `@IndexFor("#1")` as before. Because for example if seq.length = 5, index = 4, end = 5.  Then line 252 `char c1 = seq.charAt(index++);`  will be executed and cause an error( index++ is 5 and seq.charAt(5) will cause error).

File: guava/src/com/google/common/escape/UnicodeEscaper.java [hunk 5]
```diff
@@ -242,7 +254,7 @@ protected final String escapeSlow(String s, int index) {
    * @return the Unicode code point for the given index or the negated value of the trailing high
    *     surrogate character at the end of the sequence
    */
-  protected static int codePointAt(CharSequence seq, int index, int end) {
+  protected static int codePointAt(CharSequence seq, @IndexFor("#1") int index, @IndexOrHigh("#1") int end) {
```
Comment (panacekcz) [inline guava/src/com/google/common/escape/UnicodeEscaper.java]: @NhatDinh, oh, now I see. The `index` parameter should be `@IndexFor("#1")` when the method is entered, but then it can be internally increased, so it can become equal to `seq.length()`.
The problem here is that we would like the index to have type `@IndexFor("#1")` when looking at the parameter from outside the method, but `@IndexOrHigh("#1")` when looking at the variable from inside the method.
So to fix the error, I would suggest leaving `@IndexFor("#1")` on the parameter, but creating a new local variable with `@IndexOrHigh("seq")`, initializing it with the value of `index`, and replacing occurences of `index` with this new variable. That way the value of the parameter will not change in the method and the two roles (parameter vs. local variable) will be separated.

The manual (section 29.1.1) says: 
> Sometimes it is acceptable to make a refactoring, such as extracting an expression into a new local variable in order to annotate its type ...

I think we can apply it here.

Comment (panacekcz) [reply guava/src/com/google/common/escape/UnicodeEscaper.java]: We can talk about this at our next meeting.

Comment (NhatDinh) [reply guava/src/com/google/common/escape/UnicodeEscaper.java]: @panacekcz yes, I already include this problem in our next meeting Javadoc.

File: guava/src/com/google/common/escape/UnicodeEscaper.java (hunk 3 above)
Comment (panacekcz) [reply guava/src/com/google/common/escape/UnicodeEscaper.java]: @NhatDinh, do you agree with this reasoning? If yes, can you change the annotations on this method and other methods where it applies, so that we can proceed with the review?

Comment (NhatDinh) [reply guava/src/com/google/common/escape/UnicodeEscaper.java]: @panacekcz I agree with your reasoning, but like method codePointAt(), this line `index += Character.isSupplementaryCodePoint(cp) ? 2 : 1;` will cause a `compound.assignment.type.incompatible` warning. I think we need to discuss it more in our meeting.
